toolchain/base/int_store.h

```diff
@@ -0,0 +1,422 @@
+  // comparison, and so all of this ends up carefully constructed to enable very
+  // small code size when testing for an embedded value and when that test fails
+  // computing and using the index.
+  static constexpr int32_t ZeroIndexId = std::numeric_limits<int32_t>::min() >>
+                                         (TokenIdBitsShift + 1);
+
+  // The minimum embedded value in an ID.
+  static constexpr int32_t MinValue = ZeroIndexId + 1;
+
+  // The invalid ID, which needs to be placed after the largest index, which
+  // count downwards as IDs so below the smallest index ID, in order to optimize
+  // the code sequence needed to distinguish between integer and value IDs and
+  // to convert index IDs into actual indices small.
+  static constexpr int32_t InvalidId = std::numeric_limits<int32_t>::min();
+
+  // The invalid index. This is the result of converting an invalid ID into an
+  // index. We ensure that conversion can be done so that we can simplify the
+  // code that first tries to use an embedded value, then converts to an index
+  // and checks that the index is valid.
+  static const int32_t InvalidIndex;
+
+  // Document the specific values of some of these constants to help visualize
+  // how the bit patterns map from the above computations.
+  //
+  // Each bit is either `T` for part of the token or `P` as part
+  // of the available payload that we use for the ID:
+  //
+  // clang-format off: visualizing bit positions
+  //
+  //                           0bTTTT'TTTT'TPPP'PPPP'PPPP'PPPP'PPPP'PPPP
+  static_assert(MaxValue    == 0b0000'0000'0011'1111'1111'1111'1111'1111);
+  static_assert(ZeroIndexId == 0b1111'1111'1110'0000'0000'0000'0000'0000);
+  static_assert(MinValue    == 0b1111'1111'1110'0000'0000'0000'0000'0001);
+  static_assert(InvalidId   == 0b1000'0000'0000'0000'0000'0000'0000'0000);
+  // clang-format on
+
+  constexpr explicit IntId(int32_t id) : id_(id) {}
+
+  int32_t id_;
+};
+
+constexpr IntId IntId::Invalid(IntId::InvalidId);
+
+// Note that we initialize the invalid index in a constexpr context which
+// ensures there is no UB in forming it. This helps ensure all the ID -> index
+// conversions are correct because the invalid ID is at the limit of that range.
+constexpr int32_t IntId::InvalidIndex = Invalid.AsIndex();
+
+// A canonicalizing value store with deep optimizations for integers.
+//
+// This stores integers as abstract, signed mathematical integers. The bit width
+// of specific `APInt` values, either as inputs or outputs, is disregarded for
+// the purpose of canonicalization and the returned integer may use a very
+// different bit width `APInt` than was used when adding. There are also
+// optimized paths for adding integer values representable using native integer
+// types.
+//
+// Because the integers in the store are canonicalized without a specific bit
+// width there are helper functions to coerce them to a specific desired bit
```

Comment:
```suggestion
// width, there are helper functions to coerce them to a specific desired bit
```

To comment, I think it's usually "Because X, Y". The comma is because you've switched around what would directly be written as "Y because X". (Because of reversed clause order, the comma.)